Test functionalization pass in python

bdhirsh opened the pull request:
Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#66101 Test functionalization pass in python**
* #67878 functionalization: map copy_() -> to().expand_as()
* #67715 bugfix for conditional functionalization

Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

bdhirsh committed:
Make a python-level FunctionalTensor()

[ghstack-poisoned]

bdhirsh committed:
Update on "Make a python-level FunctionalTensor()"




[ghstack-poisoned]

ezyang commented:
> Where should LoggingTensor go in the longer term? Right now it's in a random test file and it's not a public API - do we eventually want it to be? Maybe for now I should just move it into some shared utils testing file so I can at least share it across tests. For now I dumped my FunctionalTensor in the same file just to get stuff running, but I need to move it somewhere else.

Tactically, stuffing it in testing/_internals is probably the easiest.

Longer term, I think LoggingTensor is going to die a fiery death once Python modes get online, since modes seem like a much more superior way to get logging, and then you'd only use the value-based version to test non-mode dispatch. As for public API, we should probably try to get the new mode to replace torch.jit.trace, but I'm not exactly sure if the old API should be preserved as is or not.

> There are still some rough edges: FunctionalTensor(FunctionalTensor(…)) breaks horrible and I haven’t looked into what breaks. I'm thinking of landing this stack before delving more into that though.

Probably level confusion? In any case, I agree, fix it later.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -369,9 +370,9 @@ static PyObject* THPVariable_make_wrapper_subclass(PyObject*, PyObject* args, Py
   // TODO: strides variant?
   static PythonArgParser parser({
-    "_make_wrapper_subclass(PyObject* cls, IntArrayRef size, *, MemoryFormat? memory_format=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
+    "_make_wrapper_subclass(PyObject* cls, IntArrayRef size, IntArrayRef strides, *, MemoryFormat? memory_format=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
```
BC-breaking! And I also agree it should be optional

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -497,5 +497,187 @@ def wrap(e):
+        return f"FunctionalTensor({self.elem._from_functional_tensor()})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
```
I'm not sure why this `__torch_dispatch__` is necessary at all. Don't we have registered kernels for the Functional dispatch key anyway; why isn't this simply a matter of just doing ordinary dispatch?

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -497,5 +497,187 @@ def wrap(e):
+                z2.add_(tmp)
+            if logging:
+                self.assertExpectedInline('\n'.join(logs), """\
+$1 = torch._ops.aten.view($0, [8])
```
Use `log_input("x", x)` to get `$0` assignment to show up in log

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -497,5 +497,187 @@ def wrap(e):
+$11 = torch._ops.aten.unsqueeze($10, 0)
+$12 = torch._ops.aten.squeeze($11)
+$13, $14 = torch._ops.aten.split($12, 2)
+$15 = torch._ops.aten.add($13, tensor([[1., 1.],
```
How come there aren't a bunch of adds for all the other intermediate quantities here?

bdhirsh commented on test/test_python_dispatch.py:
```diff
@@ -497,5 +497,187 @@ def wrap(e):
+
+    elem: torch.Tensor
+
+    __slots__ = ['elem', 'elem_']
```
relic from testing, I'll clean this up when I move this code into a new file (update coming soon)

bdhirsh committed:
Update on "Make a python-level FunctionalTensor()"

This PR adds a python-level `FunctionalTensor` subclass in python that I use to do some basic testing for the functionalization pass. I technically have a working `functionalize()` transform sitting in a feature branch in the functorch repo, but this will let me more easily test functionalization in-tree (thanks @ezyang for the idea).

It's not ready to merge yet, since there are a couple questions I should address first:
- Where should `LoggingTensor` go in the longer term? Right now it's in a random test file and it's not a public API - do we eventually want it to be? Maybe for now I should just move it into some shared utils testing file so I can at least share it across tests. For now I dumped my `FunctionalTensor` in the same file just to get stuff running, but I need to move it somewhere else.
- I'm not too sure what the state of @albanD's python-level TensorWrapper changes are - maybe I should incorporate some of those changes into this PR (specifically some of the stuff in the notebook here: https://colab.research.google.com/drive/1_nO--1y1kzDs1gcQtE8tDxyaiZh-Bl10). That's probably also related to what functorch's `BatchedTensor()` python construct should look like (cc @zou3519)
- There are still some rough edges: FunctionalTensor(FunctionalTensor(…)) breaks horrible and I haven’t looked into what breaks. I'm thinking of landing this stack before delving more into that though.

Another note on the wrapper subclass design - there are two hacks that I added in for now:
- I just co-opted `_make_wrapper_subclass` completely to get it to return a `FunctionalTensorWrapper` so my tests would pass. I should replace it with whatever's in the above TensorWrapper notebook for doing a more principled version of python<->C++ tensor wrappers.
- Currently, `_make_wrapper_subclass` doesn't the strides of its inner tensor. This is a problem for all composite operators that happen to use `self.strides()` (or `self.storage_offset()`) anywhere, because operator decomposition runs before we hit torch_dispatch. I added `strides` as a required field into that function for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "Make a python-level FunctionalTensor()"

This PR adds a python-level `FunctionalTensor` subclass in python that I use to do some basic testing for the functionalization pass. I technically have a working `functionalize()` transform sitting in a feature branch in the functorch repo, but this will let me more easily test functionalization in-tree (thanks @ezyang for the idea).

It's not ready to merge yet, since there are a couple questions I should address first:
- Where should `LoggingTensor` go in the longer term? Right now it's in a random test file and it's not a public API - do we eventually want it to be? Maybe for now I should just move it into some shared utils testing file so I can at least share it across tests. For now I dumped my `FunctionalTensor` in the same file just to get stuff running, but I need to move it somewhere else.
- I'm not too sure what the state of @albanD's python-level TensorWrapper changes are - maybe I should incorporate some of those changes into this PR (specifically some of the stuff in the notebook here: https://colab.research.google.com/drive/1_nO--1y1kzDs1gcQtE8tDxyaiZh-Bl10). That's probably also related to what functorch's `BatchedTensor()` python construct should look like (cc @zou3519)
- There are still some rough edges: FunctionalTensor(FunctionalTensor(…)) breaks horrible and I haven’t looked into what breaks. I'm thinking of landing this stack before delving more into that though.

Another note on the wrapper subclass design - there are two hacks that I added in for now:
- I just co-opted `_make_wrapper_subclass` completely to get it to return a `FunctionalTensorWrapper` so my tests would pass. I should replace it with whatever's in the above TensorWrapper notebook for doing a more principled version of python<->C++ tensor wrappers.
- Currently, `_make_wrapper_subclass` doesn't the strides of its inner tensor. This is a problem for all composite operators that happen to use `self.strides()` (or `self.storage_offset()`) anywhere, because operator decomposition runs before we hit torch_dispatch. I added `strides` as a required field into that function for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.




[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.




[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.




[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

zou3519 reviewed (commented):
This looks mostly good to me. There are two things we should probably fix here before moving on:
1. enable/disable functionalization should be wrapped in try-catch or a context manager. Otherwise, if an error happens in between them, then the TLS state will not get reset and subsequent tests in other test files will fail
2. https://github.com/pytorch/pytorch/pull/66101/files#r739325440 . A good coding practice is to separate the mechanism (the testing mechanism) from the policy (the test case, aka the function we're writing). Right now the function does things like capture logs and assert the log is correct, is it possible to make it so the function doesn't need to worry about those?

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -0,0 +1,215 @@
+        self.assertEqual(inpt, torch._from_functional_tensor(input_functional))  # input mutations should still occur
+
+    def test_simple(self):
+        def f(x, *, logging: bool):
+            # simple test: 1 view op, 1 inplace op
+            with capture_logs() as logs:
+                maybe_log_functional_input("x", x, logging)
+                tmp = torch.ones(4, 2)
+                y = x.view(4, 2)
+                y.add_(tmp)
+                z = x * x
+            if logging:
+                self.assertExpectedInline('\n'.join(logs), """\
+$0 = input('x')
+$1 = torch._ops.aten.view($0, [4, 2])
+$2 = torch._ops.aten.add($1, tensor([[1., 1.],
+        [1., 1.],
+        [1., 1.],
+        [1., 1.]]))
+$3 = torch._ops.aten.view($2, [4, 2])
+$4 = torch._ops.aten.mul($3, $3)""")
+            else:
+                return y
+        x = torch.ones(4, 2)
+        self.assert_functionalization(f, x)
```
Totally agreed. I originally wrote it that way to make sure that the expecttest output was next to the actual function for ease of comparison, but your way looks cleaner and accomplishes the same thing.

bdhirsh replied:
oh, I tried it out and I don't actually think that this works. See https://github.com/ezyang/expecttest/blob/4be7413c98b4c6bdb4b26bfc8358c0a909c8d675/expecttest/__init__.py#L197 - the expected string argument to `assertExpectedInline()` needs to be a string literal that the expecttests can update in place- otherwise it won't know where the `expected_logs` variable came from.

zou3519 replied:
I agree with your assessment, that is unfortunate. How about the following instead? We could have `assert_functionalization` return the logs:
```py
def f(x):
  tmp = torch.ones(4, 2)
  y = x.view(4, 2)
  y.add_(tmp)
  z = x * x
  return y

x = torch.ones(4, 2)
logs = self.run_functionalization_checks(f, x, return_logs=True)
self.assertExpectedInline('\n'.join(logs), """
...
""")
```

zou3519 replied:
(if that doesn't work out -- let's just ship this, we can always figure out how to make it better later...)

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -0,0 +1,215 @@
+        # y and z are aliases inside of the function, and that aliasing relationship should be maintained.
+        _y = torch._from_functional_tensor(y)
+        _z = torch._from_functional_tensor(z)
+        # We don't currently have a good way of testing aliasing relationships in python so...
```
I didn't know about the `_base` attribute, but is that supposed to work? I ran a quick test:
```
>>> import torch
>>> a = torch.ones(2)
>>> b = torch.ones(2)
>>> c = a.view(2)
>>> a._base is b._base
True # these aren't views though- should return False
>>> a._base is c._base
False # these ARE views - should return True
>>> b._base is c._base
False
```

zou3519 replied:
Oops, I gave you the wrong instructions. So here's how `_base` works:
- it returns None if there is no _base. (And `None is None` which is what is going on in your example haha)
- it returns a Tensor if there is a _base.

So here's how to write a function that checks if two things alias:
```
def are_aliased(x, y):
  if x._base is None and y._base is None:
    return False
  if x._base is not None and y._base is None:
    return x._base is y
  if x._base is None and y._base is not None:
    return y._base is x
  return x._base is y._base
```

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31942095).

bdhirsh commented on torch/testing/_internal/logging_tensor.py:
```diff
@@ -0,0 +1,109 @@
+        # The wrapping tensor (LoggingTensor) shouldn't hold any
+        # memory for the class in question, but it should still
+        # advertise the same device as before
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls, elem.size(),
+            strides=elem.stride(), storage_offset=elem.storage_offset(),
```
nope (I wish there was a way to make that clearer in the GitHub UI). I copied the contents here, and added the extra stride/storage_offset arguments.

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -0,0 +1,215 @@
+            return x
+        self.assert_functionalization(f, torch.ones(2))
+        logs = self.get_logs(f, torch.ones(2))
+        self.assertExpectedInline('\n'.join(logs), """$1 = torch._ops.aten.expand($0, [2])""")
```
I added this test for `copy_()`, but the logging tensor output looks wrong (we should be logging `aten.to` before the expand call).

I'm hoping to look at this tomorrow, but I'm wondering if it's maybe a LoggingTensor interaction problem, instead of a bug in the previous PR. I already manually tested the conditional functionalization on diag_embed (from https://github.com/pytorch/functorch/pull/235), and that seems to work - and correctly call `to` and `expand_as`.

zou3519 replied:
This seems fine to me. LoggingTensor is logging at the `__torch_dispatch__` level. What this means is that it'll only log non-CompositeImplicitAutograd operations. When `Tensor::to` is a no-op, like here, it will not emit any CompositeImplicitAutograd operations; instead it just returns the original tensor as-is. A similar thing is happening with `expand_as`: we see `expand` in the Log because expand_as is CompositeImplicitAutograd but `expand` isn't.

bdhirsh replied:
Good catch - completely forgot that if a composite kernel no-ops, LoggingTensor will not see anything.

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh committed:
Update on "Test functionalization pass in python"

Updated description:

This PR tests the functionalization pass in python in two ways. For each of the test programs that I have in `test_functionalization.py`, it:
- runs the program with and without functionalization, and asserts the outputs and (potentially mutated) inputs are equal in both cases
- runs the program with `LoggingTensor`, and uses expecttests on the resulting graph. I manually confirm that the graphs look reasonable and only contain functional ops.

Mechanically, the changes include:
- factoring out `LoggingTensor` into a testing util so it can be re-used in multiple tests
- adding some private python api's in the `torch` namespace as hooks that I can use during testing

In the original version of this PR, I also added some fixes to the `_make_subclass()` function in python: allowing you to pass in strides and storage_offset. I kept them in mainly because the changes were already there.


Differential Revision: [D31942095](https://our.internmc.facebook.com/intern/diff/D31942095)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31942095).
